### PR TITLE
Use setuptools-scm to automatically generate the version number

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

- Use setuptools-scm to automatically generate the [version number](https://github.com/pypa/setuptools_scm/#default-versioning-scheme)
- Update the readthedocs configuration so it uses a recent Python version that is compatible with this system

See https://github.com/ESMValGroup/ESMValTool/discussions/2299

Link to documentation: https://esmvaltool--1578.org.readthedocs.build/projects/ESMValCore/en/1578/contributing.html#increase-the-version-number

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by zklaus** [comment]
Looks good, like it. When doing the next conda-forge release, we will need to take this into account. Should we place a link to the following somewhere?

https://conda-forge.org/docs/maintainer/knowledge_base.html#using-setuptools-scm

**Review by znicholls — APPROVED** [review]
lgtm

File: esmvalcore/_version.py
Comment (bouweandela) [reply esmvalcore/_version.py:4]: It works and the package name is ESMValCore
https://github.com/ESMValGroup/ESMValCore/blob/8af46d466f94f4441778ed612b3b8393f7d463d8/setup.py#L194

Comment (bouweandela) [reply esmvalcore/_version.py:4]: Lowercase works too though, so I think the command is case insensitive. Probably makes sense, given that Python works on Windows too.

Comment (znicholls) [reply esmvalcore/_version.py:4]: Nice. I'm not sure what magic is happening because in other cases things are case sensitive (question for another day)

```sh
In [1]: import ESMValCore
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Input In [1], in <cell line: 1>()
----> 1 import ESMValCore

ModuleNotFoundError: No module named 'ESMValCore'

In [2]: import esmvalcore
```

**Comment by bouweandela** [comment]
Regarding code coverage: I added a test for the default case. I'm not sure how to add a test for the case where the package is somehow not installed. Anyone ideas? Shall we skip testing this since it seems unlikely to happen anyway?

**Comment by zklaus** [comment]
I agree that we don't need to test that. But thinking on how this might happen, I think one possibility is that somebody tries to import the package (for example to use a preprocessor in ipython) without it being installed, while being in the repo root, so that python imports it from the local filesystem. This might lead to surprising consequences, so I think perhaps we should not `except` the error or at least issue a warning like "Something seems wrong with your esmvaltool installation."

**Comment by bouweandela** [comment]
Does this look acceptable @zklaus?
```python
>>> import esmvalcore
/home/bandela/src/esmvalgroup/esmvalcore/esmvalcore/_version.py:8: UserWarning: No valid version information available, please install the package.
  warnings.warn(
>>> esmvalcore.__version__
'unknown'
```

**Comment by zklaus** [comment]
Yeah, I think that's fine. The average user will still be confused, but it gives enough info to receive the correct help. And it will remind me... :stuck_out_tongue_closed_eyes:

**Comment by znicholls** [comment]
> Anyone ideas?

You can mock it but I don't know actually know what leads to that particular exception being raised apart from the example that @zklaus gave... The exception that I think can occur is if someone installs a local copy but somehow doesn't have the git repo, then the version can't be inferred by setuptools-scm. I would probably either 'no cover' this branch until there's a clear example or remove the `try except` (will be very annoying for the first person who really needs it but you at least get a concrete use case then).

**Comment by zklaus** [comment]
That's a good point. @bouweandela, have you tried what happens if no git information is available?

**Comment by bouweandela** [comment]
As far as I understand, the git information is only used when you create the distribution. Once installed, it will use the version information from the installed copy.

**Comment by bouweandela** [comment]
Indeed, running `pip install .` without a .git folder crashes with a
```
LookupError: setuptools-scm was unable to detect version for /tmp/pip-req-build-o0wbzsiw.
```
importing the package without installing works, but of course without a version number:
```python
>>> import esmvalcore
/home/bandela/src/esmvalgroup/test/esmvalcore/_version.py:9: UserWarning: No version information available, please install the package.
  warnings.warn(
>>> esmvalcore.__version__
'unknown'
```

**Comment by bouweandela** [comment]
I added a fallback_version and installation now also works if no .git folder is available.

**Comment by zklaus** [comment]
Hm. Maybe I like it more without the fallback, actually. That way, it is clear something is afoot. If an expert user really wants to do that, they can set the environment variable as described in the conda-forge docs.

**Comment by bouweandela** [comment]
I found a simple way to obtain the code without the git repository: this happens if you click the green `Code` button and then `Download ZIP` from the main GitHub page of the repo. So maybe this something that inexperienced users could accidentally do.

**Comment by zklaus** [comment]
Agreed, @bouweandela. In case this wasn't clear, I still don't like the fallback in this case, since it will give us no clue whatsoever if the user then comes back for help. Perhaps we should put in another fallback, like `{next_version}.dev-detached` or something silly like that? Of course, that means that we need to update the file on every release.

**Comment by znicholls** [comment]
Sounds smart to me

**Comment by zklaus** [comment]
Alright, @bouweandela, go ahead. I think it would be nice to have the last version number as fallback, but it's not worth holding us back and maybe it's not so useful anyway. So let's do this as is.

**Comment by bouweandela** [comment]
I just saw that the error message is pretty good when trying to install without git available
```
rm -rf .git
pip install -e .
```
results in 
```
....
      LookupError: setuptools-scm was unable to detect version for /home/bandela/src/esmvalgroup/ESMValCore-main.
      
      Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.
      
      For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
 ```
so I would also be fine with just letting that error.

Importing the package without even installing is unlikely to work anyway because you're probably missing all kinds of dependencies (at least until we make something available to get all dependencies from conda). I would also be fine with letting that fail, I can make a nice error message instead of the current warning.

**Comment by bouweandela** [comment]
I think I may need to update the Dockerfile too

**Comment by bouweandela** [comment]
Ok, now cf-units seems to run into https://github.com/SciTools/cf-units/issues/218

**Comment by bouweandela** [comment]
Now I made so many changes to the documentation build that I thought it would be better to move those to a new pull request: #1588. Once that is merged, we can come back to this.

**Comment by znicholls** [comment]
I just tried this in another branch of mine, works great @bouweandela 

image

**Comment by bouweandela** [comment]
Merging this since there seem to be no objections.

**Pull request merged**
